src/planner/partialize.c

```diff
@@ -185,6 +197,652 @@ partialize_agg_paths(RelOptInfo *rel)
+	Assert(list_length(subpath->pathtarget->exprs) == list_length(parent_path->pathtarget->exprs));
+	subpath->pathtarget->sortgrouprefs = parent_path->pathtarget->sortgrouprefs;
+
+	if (can_sort)
+	{
+		*sorted_paths = lappend(*sorted_paths,
+								create_sorted_partially_agg_path(root,
+																 subpath,
+																 chunktarget,
+																 d_num_groups,
+																 extra_data));
+	}
+
+	if (can_hash)
+	{
+		*hashed_paths = lappend(*hashed_paths,
+								create_hashed_partially_agg_path(root,
+																 subpath,
+																 chunktarget,
+																 d_num_groups,
+																 extra_data));
+	}
+}
+
+/*
+ * Generate a total aggregation path for partial aggregations
+ */
+static void
+generate_agg_pushdown_path(PlannerInfo *root, Path *cheapest_total_path, RelOptInfo *output_rel,
+						   RelOptInfo *partially_grouped_rel, PathTarget *grouping_target,
+						   PathTarget *partial_grouping_target, bool can_sort, bool can_hash,
+						   double d_num_groups, GroupPathExtraData *extra_data)
+{
+	/* Get subpaths */
+	List *subpaths = get_subpaths_from_append_path(cheapest_total_path, false);
+
+	/* No subpaths available or unsupported append node */
+	if (subpaths == NIL)
+		return;
+
+	/* Generate agg paths on top of the append children */
+	List *sorted_subpaths = NIL;
+	List *hashed_subpaths = NIL;
+
+	ListCell *lc;
+	foreach (lc, subpaths)
+	{
+		Path *subpath = lfirst(lc);
+
+		/* Check if we have an append path under an append path (e.g., a partially compressed
+		 * chunk)*/
+		List *subsubpaths = get_subpaths_from_append_path(subpath, false);
+
+		if (subsubpaths != NIL)
+		{
+			List *sorted_subsubpaths = NIL;
+			List *hashed_subsubpaths = NIL;
+
+			ListCell *lc2;
+			foreach (lc2, subsubpaths)
+			{
+				Path *subsubpath = lfirst(lc2);
+
+				add_partially_aggregated_subpaths(root,
+												  cheapest_total_path,
+												  partial_grouping_target,
+												  d_num_groups,
+												  extra_data,
+												  can_sort,
+												  can_hash,
+												  subsubpath,
+												  &sorted_subsubpaths /* Result path */,
+												  &hashed_subsubpaths /* Result path */);
+			}
+
+			if (can_sort)
+			{
+				sorted_subpaths = lappend(sorted_subpaths,
+										  copy_append_like_path(root, subpath, sorted_subsubpaths));
+			}
+
+			if (can_hash)
+			{
+				hashed_subpaths = lappend(hashed_subpaths,
+										  copy_append_like_path(root, subpath, hashed_subsubpaths));
+			}
+		}
+		else
+		{
+			add_partially_aggregated_subpaths(root,
+											  cheapest_total_path,
+											  partial_grouping_target,
+											  d_num_groups,
+											  extra_data,
+											  can_sort,
+											  can_hash,
+											  subpath,
+											  &sorted_subpaths /* Result paths */,
+											  &hashed_subpaths /* Result paths */);
+		}
+	}
+
+	/* Create new append paths */
+	cheapest_total_path->pathtarget = partial_grouping_target;
```

Comment:
Is it OK/safe/best practice to modify the pathtarget of an existing path like this to turn it into a partial path? Wondering if we shouldn't simply add partial paths, in addition to existing paths, then run `set_cheapest` and use cost to figure out which on is the best.

Reply:
I changed the functionality of `copy_append_like_path`. So, the `pathtarget` is only changed on the copied path.